### ISSUE TITLE
Remove or Reorder Objects in An Array

### ISSUE THREAD

**Issue by dafeder** [GetDKAN/dkan#4335]
More discovery work is needed to determine the approach we want to take to solve this.

--

If there is an array in the schema you can only add and remove the last item in the dataset form.

For instance, if there are 3 distributions, you cannot change their order or remove just the first or middle one. 

A user should be able to remove or reorder any object in the list

**Comment by dafeder** [comment]
Getting close on something custom here with all buttons and fieldsets, though starting to think that, at least with simple text fields, something draggable w/tables would be better. See https://api.drupal.org/api/examples/modules%21tabledrag_example%21tabledrag_example.module/group/tabledrag_example/4.0.x

**Comment by stefan-korn** [comment]
@dafeder : I tested the changes in my setup with different schema and came across a few issues:
- does not work correct on subschema, i. e. you have a multivalue field in distribution and either edit this field in the dataset form or in the distribution form. The actions buttons won't work (mismatch of ajax target wrapper)
- if you have a simple text element and want to use textarea widget on that. This will also not work correct. This is probably because `buildSimpleArrayElement` is adding a `field` sub key to store the widget type "textfield". textarea widget changes this widget type to "textarea" and adds some more textarea stuff, but textarea widget expects this to live at the root of the array, where now the "fieldset" type resides.

Fiddling a bit with the code I could make it work somehow in my setup. But this seems very hacky at the moment, so I am not sure whether this would be suitable to solve it more generally.

What is your opinion on this feature currently? Do you see it as ready for use in a default setup of DKANv2 (default schema)? Or do you also have points that you still want to address?

In terms of user experience this would be kind of important for us. We will have datasets with several distributions imported from DKANv1. Now if you later want to remove a distribution, this gets difficult without this feature.

**Comment by dafeder** [comment]
@stefan-korn yeah it's very frustrating, I spent so long getting that PR to the place it was and it still had so many bugs and edge cases I needed to put it aside and haven't gotten back to it. I agree this is important functionality, I worked on another site that had a smaller object array (for author fields) that really suffered from not being able to change the order. Did your changes fix the issue with the file upload widget? That was the biggest hurdle for me. I feel like this must be solvable, but yeah every time an edge case is fixed it adds more complexity that introduces new bugs 😦 

We'll take a look at your code asap. Maybe we need to rewrite some of this from higher up, the order of operations between the different classes could maybe be simplified. IE maybe we should be merging the two JSON schemas before starting to build the form, so that we don't have to iterate through the form elements multiple times and change things that are already there.

Another thing we've been playing with for data dictionaries has been to not nest a whole form for an object but to just have a summary table that you can edit individual rows of in a modal... that may not work for multiple levels of nesting but does take some pressure off the form API.

**Comment by stefan-korn** [comment]
@dafeder : Thanks for your reply.

Please don't spend much time looking into my code. I thought I fixed a few things, but now I am recognizing that I may have broken others. So this code in my PR is really far from usable.

Regarding file upload widget, I am not sure what the problem is you were experiencing. Just uploading a file seem to work fine.
There were some problems with file upload that we could resolve with #3611 . Not sure if this maybe overlaps here.

I was also loosely thinking about a different approach, maybe as a workaround. I thought whether to have something like a checkbox to remove items and do the removal on saving the form. Though this would not help for sorting, but our case is probably more about deleting than sorting. The benefit from that, would be that one would not need to change the structure (the fieldset container) and would not need that ajax targeting. But it does also not feel like a nice solution.

**Comment by stefan-korn** [comment]
@dafeder : I have implemented a solution to remove items via the module https://www.drupal.org/project/dkan_json_form_tweak (latest commit) - It works by adding a "remove" checkbox and then removing based on value of checkbox during the value handler processing.

It works for arrays of objects in dataset form and subschema forms. I will try to further improve it, but for the moment it cures our biggest pain on removing items.

**Comment by TheETupper** [comment]
@kaise-lafrai  took a pass at using different modules and field type. @dafeder is gonna take another pass. The file id is a hidden field on the form. Only when you add a new file once saved does this disappears.

**Comment by dafeder** [comment]
Dependent on work being done #4461

**Comment by dafeder** [comment]
Still working on this, though may set it aside for a while again. See PR for latest issues.
